spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/pubsub/it/PubSubChannelAdaptersIT.java

```diff
@@ -0,0 +1,231 @@
+	}
+
+	@Test
+	public void sendAndReceiveMessage() {
+		Map<String, Object> headers = new HashMap<>();
+		// Only String values for now..
+		headers.put("storm", "lift your skinny fists");
+		headers.put("static", "lift your skinny fists");
+		headers.put("sleep", "lift your skinny fists");
+
+		this.inputChannel.send(
+				MessageBuilder.createMessage("I am a message.",  new MessageHeaders(headers)));
+
+		Message<?> message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		assertThat(message.getPayload()).isInstanceOf(String.class);
+		String payload = (String) message.getPayload();
+		assertThat(payload).isEqualTo("I am a message.");
+
+		assertThat(message.getHeaders().size()).isEqualTo(6);
+		assertThat(message.getHeaders().get("storm")).isEqualTo("lift your skinny fists");
+		assertThat(message.getHeaders().get("static")).isEqualTo("lift your skinny fists");
+		assertThat(message.getHeaders().get("sleep")).isEqualTo("lift your skinny fists");
+	}
+
+	@Test
+	@DirtiesContext
+	public void sendAndReceiveMessageInBytes() {
+		this.inboundChannelAdapter.setMessageConverter(null);
+		this.inputChannel.send(MessageBuilder.withPayload("I am a message.").build());
+
+		Message<?> message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		assertThat(message.getPayload()).isInstanceOf(byte[].class);
+		String stringPayload = new String((byte[]) message.getPayload());
+		assertThat(stringPayload).isEqualTo("I am a message.");
+	}
+
+	@Test
+	@DirtiesContext
+	public void sendAndReceiveMessageManualAck() {
+		this.inboundChannelAdapter.setAckMode(AckMode.MANUAL);
+		this.inputChannel.send(MessageBuilder.withPayload("I am a message.").build());
+
+		Message<?> message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		AckReplyConsumer acker =
+				(AckReplyConsumer) message.getHeaders().get(GcpHeaders.ACKNOWLEDGEMENT);
+		assertThat(acker).isNotNull();
+		acker.nack();
+		message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		acker = (AckReplyConsumer) message.getHeaders().get(GcpHeaders.ACKNOWLEDGEMENT);
+		assertThat(acker).isNotNull();
+		acker.ack();
+		message = this.channel.receive(5000);
+		assertThat(message).isNull();
+	}
+
+	@Test
+	@DirtiesContext
+	public void sendAndReceiveMessagePublishCallback() {
+		ListenableFutureCallback callbackSpy = Mockito.spy(new ListenableFutureCallback<String>() {
+			@Override
+			public void onFailure(Throwable ex) {
+
+			}
+
+			@Override
+			public void onSuccess(String result) {
+
+			}
+		});
+		this.outboundChannelAdapter.setPublishCallback(callbackSpy);
+		this.inputChannel.send(MessageBuilder.withPayload("I am a message.").build());
+
+		Message<?> message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		verify(callbackSpy, times(1)).onSuccess(any());
+	}
+
+	@Configuration
+	@EnableIntegration
+	static class IntegrationConfiguration {
+
+		@Autowired
+		@Lazy
```

Comment:
Why `@Lazy` ?

Reply:
Because there is a theoretical circular dependency.

`IntegrationConfiguration` requires a `PubSubTemplate` bean that is created by `GcpPubSubAutoConfiguration`.
`GcpPubSubAutoConfiguration` requires a `GcpProjectIdProvider` bean that is provided by `IntegrationConfiguration` in this case.

The best way I found of resolving this issue was to make `PubSubTemplate` lazy, so that it's only instantiated when we need it. When that happens, the `GcpProjectIdProvider` is already in the context, so we're all good.

Reply:
Yes, good point!

The easy fix is to make that `gcpProjectIdProvider()` method as `static`. 
Another way is to use method argument injection instead of this field.

Reply:
Actually, getting rid of `GcpContextAutoConfiguration` resolved the circular dependency for us.